gulpfile.js/changelog.js

```diff
@@ -0,0 +1,380 @@
+const git = require('simple-git/promise')(__dirname);
+
+const { changelog } = require('./paths');
+
+
+const ISSUE_RE = /#(\d+)(?![\d\]])/g;
+const ISSUE_SUB = '[#$1](https://github.com/PrismJS/prism/issues/$1)';
+
+function linkify(cb) {
+	return pump([
+		src(changelog),
+		replace(ISSUE_RE, ISSUE_SUB),
+		replace(
+			/\[[\da-f]+(?:, *[\da-f]+)*\]/g,
+			m => m.replace(/([\da-f]{7})[\da-f]*/g, '[`$1`](https://github.com/PrismJS/prism/commit/$1)')
+		),
+		dest('.')
+	], cb);
+}
+
+/**
+ * Creates an array which iterates its items in the order given by `compareFn`.
+ *
+ * The array may not be sorted at all times.
+ *
+ * @param {(a: T, b: T) => number} compareFn
+ * @returns {T[]}
+ * @template T
+ */
+function createSortedArray(compareFn) {
+	const a = [];
+
+	a['sort'] = function () {
+		return Array.prototype.sort.call(this, compareFn);
+	};
+	a[Symbol.iterator] = function* () {
+		const copy = [];
+		for (let i = 0; i < this.length; i++) {
+			copy.push(this[i]);
+		}
+		copy.sort(compareFn);
+		for (const item of copy) {
+			yield item;
+		}
+	};
+
+	return a;
+}
+
+/**
+ * Parses the given log line and adds the list of the changed files to the output.
+ *
+ * @param {string} line A one-liner log line consisting of the commit hash and the commit message.
+ * @returns {Promise<CommitInfo>}
+ *
+ * @typedef {{ message: string, hash: string, changes: CommitChange[] }} CommitInfo
+ * @typedef {{ file: string, mode: ChangeMode }} CommitChange
+ * @typedef {"A" | "C" | "D" | "M" | "R" | "T" | "U" | "X" | "B"} ChangeMode
+ */
+function getCommitInfo(line) {
+	const [, hash, message] = /^([a-f\d]+)\s+(.*)$/i.exec(line);
+
+	return git.raw(['diff-tree', '--no-commit-id', '--name-status', '-r', hash]).then(res => {
+		/* The output looks like this:
+		 *
+		 * M       components.js
+		 * M       components.json
+		 *
+		 * or nothing for e.g. reverts.
+		 */
+
+		const changes = !res ? [] : res.trim().split(/\n/g).map(line => {
+			const [, mode, file] = /(\w)\s+(.+)/.exec(line);
+			return { mode: /** @type {ChangeMode} */ (mode), file };
+		});
+		return { hash, message, changes };
+	});
+}
+
+/**
+ * Parses the output of `git log` with the given revision range.
+ *
+ * @param {string | Promise<string>} range The revision range in which the log will be parsed.
+ * @returns {Promise<CommitInfo[]>}
+ */
+function getLog(range) {
+	return Promise.resolve(range).then(range => {
+		return git.raw(['log', range, '--oneline']);
+	}).then(res => {
+		/* The output looks like this:
+		 *
+		 * bfbe4464 Invoke `callback` after `after-highlight` hook (#1588)
+		 * b41fb8f1 Fixes regex for JS examples (#1591)
+		 */
+		const commits = !res ? [] : res.trim().split(/\n/g);
+
+		return Promise.all(commits.map(getCommitInfo));
+	});
+}
+
+const revisionRanges = {
+	nextRelease: git.raw(['describe', '--abbrev=0', '--tags']).then(res => `${res.trim()}..HEAD`),
+	v1_15_0__HEAD: 'v1.15.0..HEAD',
+	v1_14_0__v1_15_0: 'v1.14.0..v1.15.0',
```

Comment:
Does this need to change every release? What's the story here?

Reply:
Just some constants I used during development.
I'll remove them.